src/main/java/org/apache/commons/text/CasedString.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.commons.text;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * Handles converting from one string case to another (e.g. camel case to snake case).
+ * @since 1.13.0
+ */
+public class CasedString {
+    /** the string the the cased format. */
+    private final String string;
+    /** the case of the string. */
+    private final StringCase stringCase;
+
+    /**
+     * A method to join camel string fragments together.
+     */
+    private static final Function<String[], String> CAMEL_JOINER = a -> {
+        StringBuilder sb = new StringBuilder(a[0]);
+
+        for (int i = 1; i < a.length; i++) {
+            sb.append(WordUtils.capitalize(a[i]));
+        }
+        return sb.toString();
+    };
+
+    /**
+     * An enumeration of supported string cases.
+     */
+    public enum StringCase {
+        /**
+         * Camel case identifies strings like 'CamelCase'.
```

Comment:
Each string case should have a precise definition (a RegEx for example) or a link to a precise definition.